extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcServerProcessor.java

```diff
@@ -193,13 +191,30 @@ void discoverBindableServices(BuildProducer<BindableServiceBuildItem> bindables,
                 continue;
             }
             BindableServiceBuildItem item = new BindableServiceBuildItem(service.name());
+            Set<MethodInfo> blockingMethods = gatherBlockingMethods(service);
+            for (MethodInfo method : blockingMethods) {
+                item.registerBlockingMethod(method.name());
+            }
+            bindables.produce(item);
+        }
+    }
+
+    private Set<MethodInfo> gatherBlockingMethods(ClassInfo service) {
+        Set<MethodInfo> result = new HashSet<>();
+        if (service.classAnnotation(GrpcDotNames.BLOCKING) != null) {
             for (MethodInfo method : service.methods()) {
-                if (method.hasAnnotation(GrpcDotNames.BLOCKING)) {
-                    item.registerBlockingMethod(method.name());
+                if (!BLOCKING_SKIPPED_METHODS.contains(method.name())
+                        && !method.hasAnnotation(GrpcDotNames.NON_BLOCKING)) {
+                    result.add(method);
                 }
             }
-            bindables.produce(item);
         }
+        for (MethodInfo method : service.methods()) {
```

Comment:
Do we have to iterate over `service.methods()` twice?

Reply:
nope, it's either a bit more complex if or two iterations